demos/python_demos/common/images_capture.py

```diff
@@ -0,0 +1,166 @@
+            if image is not None:
+                return
+        raise OpenError("Can't read the first image from {}".format(input))
+
+    def read(self):
+        while self.file_id < len(self.names):
+            filename = os.path.join(self.dir, self.names[self.file_id])
+            image = cv2.imread(filename, cv2.IMREAD_COLOR)
+            self.file_id += 1
+            if image is not None:
+                return image
+        if self.loop:
+            self.file_id = 0
+            while self.file_id < len(self.names):
+                filename = os.path.join(self.dir, self.names[self.file_id])
+                image = cv2.imread(filename, cv2.IMREAD_COLOR)
+                self.file_id += 1
+                if image is not None:
+                    return image
+        return None
+
+    def get_resolution(self):
+        for name in self.names:
+            filename = os.path.join(self.dir, name)
+            image = cv2.imread(filename, cv2.IMREAD_COLOR)
+            if image is not None:
+                height, width, _ = image.shape
+                return (width, height)
+
+
+class VideoCapWrapper(ImagesCapture):
+
+    def __init__(self, input, loop, camera_resolution):
+        super().__init__(loop)
+        self.cap = cv2.VideoCapture()
+        try:
+            status = self.cap.open(int(input))
+            self.type = 'CAMERA'
+            self.cap.set(cv2.CAP_PROP_BUFFERSIZE, 1)
+            self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, camera_resolution[0])
+            self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, camera_resolution[1])
+            self.cap.set(cv2.CAP_PROP_AUTOFOCUS, 1)
+            self.cap.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc(*'MJPG'))
+        except ValueError:
+            status = self.cap.open(input)
+            self.type = 'VIDEO'
+        if not status:
+            raise InvalidInput("Can't open the {} from {}".format(self.get_type().lower(), input))
+
+    def read(self):
+        status, image = self.cap.read()
+        if not status:
+            if not self.loop:
+                return None
+            self.cap.set(cv2.CAP_PROP_POS_FRAMES, 0)
+            status, image = self.cap.read()
+            if not status:
+                return None
+        return image
+
+    def get_fps(self):
+        return self.cap.get(cv2.CAP_PROP_FPS)
+
+    def get_resolution(self):
+        return (int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH)),
+                int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT)))
+
+
+def open_images_capture(input, loop, camera_resolution=(1280, 720)):
+    errors = {'InvalidInput': [], 'OpenError': []}
+    for idx, reader in enumerate((DirReader, ImreadWrapper, VideoCapWrapper)):
+        try:
+            if idx != 2:
+                return reader(input, loop)
+            return reader(input, loop, camera_resolution)
+        except (InvalidInput, OpenError) as e:
+            errors[str(e)].append(e.message)
```

Comment:
Use `type(e)` as a key after updating `errors = {InvalidInput: [], OpenError: []}`. And you can append exceptions themselves.
`Exception` is derived from `BaseException` which is already good at handling `str()`.  So you don't need to define `__str__()`, print stored exceptions.